File: js/src/sha256-merkle-tree.ts
import { sha256 } from '@noble/hashes/sha256';

export class Sha256MerkleTree implements MerkleTree {

  constructor() {
  }

  getRoot(leafs: Uint8Array[]): Uint8Array {
    // Hash all the leafs first.
    let mutableLeafs = leafs.map(leaf => sha256(leaf));

    // Make sure to at least hash once even if you only start with one leaf.
    if (mutableLeafs.length % 2 !== 0) {
      mutableLeafs.push(mutableLeafs[mutableLeafs.length - 1]);
    }

    while (mutableLeafs.length > 1) {
      if (mutableLeafs.length % 2 !== 0) {
        mutableLeafs.push(mutableLeafs[mutableLeafs.length - 1]);
      }

      const newLeaves = Array.from({ length: mutableLeafs.length / 2 }, (_, i) => {
        const combinedData = Buffer.concat([mutableLeafs[i * 2], mutableLeafs[i * 2 + 1]]);

        return sha256(combinedData);
      });

      mutableLeafs = newLeaves;
    }

    return mutableLeafs[0];
  }
}

File: server/src/main/kotlin/de/tfelix/evmbitstream/bitstream/Sha256MerkleTree.kt
package de.tfelix.evmbitstream.bitstream

import org.springframework.stereotype.Component
import java.security.MessageDigest

@Component
class Sha256MerkleTree : MerkleTree {

    private val hasher = MessageDigest.getInstance("SHA-256")

    override fun getRoot(leafs: List<ByteArray>): ByteArray {
        // Hash all the leafs first.
        val hashedLeafs = leafs.map { leaf -> hasher.digest(leaf) }

        var mutableLeafs = hashedLeafs.toMutableList()

        // Make sure to at least hash once even if you only start with one leaf.
        if (mutableLeafs.size % 2 != 0) {
            mutableLeafs.add(mutableLeafs.last())
        }

        while (mutableLeafs.size > 1) {
            if (mutableLeafs.size % 2 != 0) {
                mutableLeafs.add(mutableLeafs.last())
            }

            val newLeaves = (mutableLeafs.indices step 2).map { i ->
                val combinedData = mutableLeafs[i] + mutableLeafs[i + 1]
                hasher.digest(combinedData)
            }

            mutableLeafs = newLeaves.toMutableList()
        }

        return mutableLeafs[0]
    }
}